src/ripple/overlay/Slot.h

```diff
@@ -0,0 +1,607 @@
+
+    /** Check if peers stopped relaying messages
+     * @param f Function is called for every peer in Squelched state
+     *     with arguments:
+     *     peer weak_ptr
+     */
+    template <typename F>
+    void
+    checkIdle(F&& f);
+
+private:
+    /** Reset counts of peers in Selected or Counting state */
+    void
+    resetCounts();
+
+    /** Initialize slot to Counting state */
+    void
+    initCounting();
+
+    /** Data maintained for each peer */
+    struct PeerInfo
+    {
+        std::weak_ptr<Peer> peer_;  // peer's weak_ptr
+        PeerState state_;           // peer's state
+        std::size_t count_;         // message count
+        time_point expire_;         // squelch expiration time
+        time_point lastMessage_;    // time last message received
+    };
+    std::unordered_map<id_t, PeerInfo> peers_;  // peer's data
+    std::unordered_set<id_t> considered_;       // pool of peers considered
+                                                // as the source of messages
+                                                // from validator
+    typename clock_type::time_point
+        lastSelected_;  // last time peers were selected,
+                        // used to age the slot
+    SlotState state_;   // slot's state
+};
+
+template <typename Peer, typename clock_type>
+template <typename F>
+void
+Slot<Peer, clock_type>::checkIdle(F&& f)
+{
+    auto now = clock_type::now();
+    for (auto it = peers_.begin(); it != peers_.end();)
+    {
+        auto& peer = it->second;
+        auto id = it->first;
+        ++it;
+        if (now - peer.lastMessage_ > IDLED)
+            deletePeer(id, false, std::forward<F>(f));
+    }
+}
+
+template <typename Peer, typename clock_type>
+template <typename F>
+void
+Slot<Peer, clock_type>::update(
+    id_t const& id,
+    std::weak_ptr<Peer> peerPtr,
+    protocol::MessageType type,
+    F&& f)
+{
+    auto now = clock_type::now();
+    auto it = peers_.find(id);
+    // First message from this peer
+    if (it == peers_.end())
+    {
+        peers_.emplace(std::make_pair(
+            id,
+            std::move(PeerInfo{peerPtr, PeerState::Counting, 0, now, now})));
+        initCounting();
+        return;
+    }
+    // Message from a peer with expired squelch
+    else if (
+        it->second.state_ == PeerState::Squelched && now > it->second.expire_)
+    {
+        it->second.state_ = PeerState::Counting;
+        it->second.lastMessage_ = now;
+        initCounting();
+        return;
+    }
+
+    auto& peer = it->second;
+    peer.lastMessage_ = now;
+
+    if (state_ != SlotState::Counting || peer.state_ == PeerState::Squelched)
+        return;
+
+    if (++peer.count_ > MESSAGE_THRESHOLD)
+        considered_.insert(id);
+
+    if (now - lastSelected_ > 2 * MAX_UNSQUELCH_EXPIRE)
+    {
+        initCounting();
+        return;
+    }
+
+    if (
+        considered_.size() > MAX_SELECTED_PEERS &&
+        considered_.size() >= 3 * peers_.size() / 4)
```

Comment:
discussed with @gregtatcam 
Checking `considered_.size() >= 3 * peers_.size() / 4` could delay peer selection and result in a saw tooth traffic pattern with high point being 75%.

Reply:
Changed the logic to start the selection when MAX_SELECTED_PEERS reach MAX_MESSAGE_THRESHOLD (20) and include into the selection pool all peers that reached MIN_MESSAGE_THRESHOLD (17).